### PR TITLE
GLM-5.1 prefill integration

### PR DESCRIPTION
**Pull request by nmilicevicTT**
Addresses https://github.com/tenstorrent/tt-metal/issues/47913

## What
Adds GLM-5.1 (HF `glm_moe_dsa`: absorbed-MQA MLA + DSA lightning-indexer top-2048 +
256-expert top-8 MoE, 78 layers) to the DeepSeek prefill demo, mirroring the Kimi K2.6
integration, on Blackhole Galaxy (8x4)

### CI coverage (now mirrors Kimi's galaxy triad)

| Level | Kimi | GLM (this PR) | Notes |
|---|---|---|---|
| MLA (op) | `kimi_mla` (`test_mla`) | `dsa_mla_glm` (`test_sparse_mla`) | GLM MLA is sparse DSA → lives in `sparse_mla`, not `test_mla` |
| MoE (op) | `kimi_moe` (`test_ttnn_moe`) | `glm_moe` (`test_ttnn_moe::test_ds_moe`, glm-256) | 256-expert noaux_tc MoE |
| **Block** | `kimi_prefill_block` | **`glm_prefill_block`** (`test_prefill_block.py::test_glm_prefill_block`) | **new**; see design note |

Loudbox: `demo_sp` + `blackhole_e2e` run GLM sparse-MLA `2x4` and GLM MoE `linear-8`.
Budget: `models.demo.bh_galaxy` 450 → 480.

### Block test — design note for reviewers
GLM has no runnable HF `reference_model_cls`, so `test_glm_prefill_block` can't use `run_model`
(whose golden is built from `reference_model_cls`). Instead it **composes the CPU references GLM
already owns** — `cpu_deepseek_v32.SparseMLAReference` (DSA-MLA) + a per-expert noaux_tc MoE
(`reference/glm_5_1/`) — which is exactly `TtPrefillBlock.forward`. Weights: LOAD real weights from
the prebuilt ttnn cache when present (device) + matching host weights from the checkpoint
(reference), else random; never rebuilds the cache. There is `GlmMoeDsaModel` in transformers >= 5.10 but it uses dense attn to simulate sparse and PCC is worse than our MLA CPU reference. 

### Validation (8x4, seq5120)
- `glm_prefill_block`: pretrained-from-cache **dense 0.99999996 / moe 0.99999964**;
  random **dense 0.9997 / moe 0.9946**.
- MoE reference == HF `GlmMoeDsaMoE`: **PCC 1.0** (numeric + source).
- MLA reference (`cpu_deepseek_v32`) validated vs device at **0.9953** (op-level);
  router/expert math matches HF by source inspection (full HF DSA-attention numeric
  cross-check not yet run).
- `dsa_mla_glm` 3 passed, `glm_moe` 4 passed (PCC 0.9877) on the rebased binary.

### Also in this branch
- fix: restore the `q_lora_rank` gate in `_resolve_mm_cfg` (lost in a rebase; GLM was picking up
  Kimi's overflowing q_a_proj 640 matmul config → `Num output blocks along x (13)` TT_FATAL).
- review cleanups + corrected stale `glm_moe_dsa` comments (it IS AutoConfig-loadable in
  transformers >=5.10; we compose CPU refs deliberately).

### Known follow-ups (not blockers)
- Device gate stores `e_score_correction_bias` in **bf16** even in DEVICE_FP32; GLM's real ~34.5
  bias has a +-0.14 spread finer than bf16 resolution and dominates top-8 selection — worth a
  device-gate precision fix (also relevant to the noaux_tc routing sensitivity).

## Not in CI (by design, matches Kimi/DSA family)
No transformer/chunked CI stages (transformer tests validate vs the vLLM 55k golden trace locally).

exabox env vars:
```
export GLM51_HF_MODEL=/mnt/models/zai-org/GLM-5.1
export TT_GLM51_PREFILL_TTNN_CACHE=/mnt/models/deepseek-prefill-cache/GLM-5_1-Cache
export PREFILL_TRACE_DIR=/mnt/models/deepseek-prefill-cache/golden/structured_traces/glm_51_code_debug_55k_vllm
export GLM51_MLA_REF_CACHE=/mnt/models/deepseek-prefill-cache/glm_5_1_mla_ref_cache
```

While CI is flaky: 
image

[BH e2e pipeline](https://github.com/tenstorrent/tt-metal/actions/runs/28812706033) PASS

### CI Status
_Auto-generated on every push. Badges update live. Click a badge to filter runs by this branch._

- [](https://github.com/tenstorrent/tt-metal/actions/workflows/sanity-tests.yaml?query=branch:nmilicevic/glm-integration)
- [](https://github.com/tenstorrent/tt-metal/actions/workflows/runtime-sanity-tests.yaml?query=branch:nmilicevic/glm-integration)
- [](https://github.com/tenstorrent/tt-metal/actions/workflows/blackhole-sanity-tests.yaml?query=branch:nmilicevic/glm-integration)
- [](https://github.com/tenstorrent/tt-metal/actions/workflows/tt-metal-l2-nightly.yaml?query=branch:nmilicevic/glm-integration)
- [](https://github.com/tenstorrent/tt-metal/actions/workflows/all-model-tests.yaml?query=branch:nmilicevic/glm-integration)
- [](https://github.com/tenstorrent/tt-metal/actions/workflows/pipeline-select.yaml?query=branch:nmilicevic/glm-integration)
- [](https://github.com/tenstorrent/tt-metal/actions/workflows/pipeline-select-t3k.yaml?query=branch:nmilicevic/glm-integration)
- [](https://github.com/tenstorrent/tt-metal/actions/workflows/pipeline-select-galaxy.yaml?query=branch:nmilicevic/glm-integration)

### CODE REVIEW

**Review by pavlepopovic — APPROVED** [review]

File: models/demos/deepseek_v3_d_p/tests/test_prefill_block.py [hunk 1]
```diff
@@ -719,3 +726,239 @@ def test_kimi_prefill_block(
...
+        }
+
+    gate_weights = {
+        "weight": (torch.randn(n_routed, hidden, generator=g) * hs).to(torch.bfloat16),
+        "e_score_correction_bias": (torch.randn(n_routed, generator=g) * 0.01).to(torch.float32),
+    }
+    return gate_weights, [_expert() for _ in range(n_routed)], _expert()
+
+
+def _glm_pretrained_weights(config, model_dir, layer_idx, is_moe):
```
Comment (ipotkonjak-tt) [inline models/demos/deepseek_v3_d_p/tests/test_prefill_block.py:785]: [nit] should the be in an util file? I see mainly tests and run_model above

File: models/demos/deepseek_v3_d_p/tests/test_prefill_block.py [hunk 2]
```diff
@@ -719,3 +726,239 @@ def test_kimi_prefill_block(
...
+    indirect=["mesh_device", "device_params"],
+)
+@pytest.mark.parametrize("seq_len", [5120], ids=["seq5120"])
+@pytest.mark.parametrize(
+    "layer_type, layer_idx", [("dense", 0), ("moe", GLM51Config.NUM_DENSE_LAYERS)], ids=["dense", "moe"]
+)
+@pytest.mark.parametrize("variant", ["glm_5_1"], indirect=True, ids=["glm"])
+@pytest.mark.skipif(not is_blackhole(), reason="DSA ops (indexer / sparse SDPA) are Blackhole-only")
+@pytest.mark.timeout(0)
+def test_glm_prefill_block(
```
Comment (ipotkonjak-tt) [inline models/demos/deepseek_v3_d_p/tests/test_prefill_block.py:852]: any reason we couldn't use the run_model function? I'm fine to merge and clean up later, but we should figure out if our functions are general enough, if not whether we can make the be or why not.

File: models/demos/deepseek_v3_d_p/tests/test_prefill_transformer.py [hunk 3]
```diff
@@ -1103,3 +1115,107 @@ def test_kimi_prefill_transformer(
         tokenizer,
         request,
     )
+
+
+@pytest.mark.skipif(not is_blackhole(), reason="GLM-5.1 requires Blackhole")
+@pytest.mark.parametrize("tokenizer", ["right", "left"], indirect=True, ids=["right_pad", "left_pad"])
+@pytest.mark.parametrize("temperature", [[0.5]], ids=["temp_sweep"])
```
Comment (ipotkonjak-tt) [inline models/demos/deepseek_v3_d_p/tests/test_prefill_transformer.py:1133]: i don't remember temp sweep being done for other models. Is this GLM specific? Could this be impacting our pcc curve on this model?

But I guess this is one shot path, which we are not currently investigating pcc on

File: models/demos/deepseek_v3_d_p/tt/mla/mla.py [hunk 4]
```diff
@@ -515,6 +515,12 @@ def _resolve_mm_cfg(self, weight_name: str, seq_len_local: int) -> dict | None:
         # variant at the same seq_len_local doesn't pick up a dimensionally-invalid program_config.
         if cfg is not None and cfg.get("num_heads") not in (None, self.num_heads):
             cfg = None
+        # Some of those configs are additionally q_lora_rank-specific: the 640 set's program_configs are
+        # dimensionally valid at Kimi's q_lora_rank (1536) but overflow the grid at GLM-5.1's (2048), even
+        # though both have 64 heads. When a config declares a q_lora_rank that doesn't match this model,
+        # fall back so a same-heads/same-seq variant doesn't pick up an invalid program_config.
+        if cfg is not None and cfg.get("q_lora_rank") not in (None, self.q_lora_rank):
```
Comment (ipotkonjak-tt) [inline models/demos/deepseek_v3_d_p/tt/mla/mla.py:529]: couldn't we deal with this in the config class?

**Review by ipotkonjak-tt — APPROVED** [review]
Great, thank you! There is a just few nit comments

File: models/demos/deepseek_v3_d_p/tests/test_prefill_transformer.py (hunk 3 above)
Comment (nmilicevicTT) [reply models/demos/deepseek_v3_d_p/tests/test_prefill_transformer.py:1133]: All 3 models have it, although `temp_sweep` might be a misnomer. It only affects first-token sampling; PCC is computed on raw logits/decoder outputs (pre-sampling), so it can't move the PCC curve.

File: models/demos/deepseek_v3_d_p/tt/mla/mla.py (hunk 4 above)
Comment (nmilicevicTT) [reply models/demos/deepseek_v3_d_p/tt/mla/mla.py:529]: I think this `g_lora_rank` guard is the same as `num_heads` above, used to skip a tuned config that doesn't suit us. If we put this check anywhere else it would split this decision across two files.

File: models/demos/deepseek_v3_d_p/tests/test_prefill_block.py (hunk 2 above)
Comment (nmilicevicTT) [reply models/demos/deepseek_v3_d_p/tests/test_prefill_block.py:852]: `run_model'`s path is built around `create_hf_model`() + a single HF module; GLM's only full HF module (`GlmMoeDsaModel`) is non-absorbed (256-expert, dense attention) ~15h per block, vs ~7s/block for the absorbed CPU composition that we use. This way we also use building blocks we already have MLACPU + TorchMoe. I would also prefer to use HF ref and will check if this somehow can be refactored, maybe open an issue for it.

File: models/demos/deepseek_v3_d_p/tests/test_prefill_block.py (hunk 1 above)
Comment (nmilicevicTT) [reply models/demos/deepseek_v3_d_p/tests/test_prefill_block.py:785]: The `_glm_*_weights` helpers are single-use, specific to this one block test. Kept local so the test is self-contained; a util file only pays off once they're shared

**Review by dpopovTT — APPROVED** [review]

File: .github/time_budget.yaml [hunk 5]
```diff
@@ -346,7 +346,7 @@ models:
     wh_llmbox_perf: 1290
     wh_galaxy: 345
     wh_galaxy_perf: 150
-    bh_galaxy: 410
+    bh_galaxy: 480
```
Comment (roseli-TT) [inline .github/time_budget.yaml]: This is a sizable increase to bh galaxy time in CI. Is there any way this can be reduced?

Comment (nmilicevicTT) [reply .github/time_budget.yaml]: Since the galaxy CI is a bit flaky, I rerun the added tests locally, measured and lowered the budget to 440 -> a 30min increase (5 + 20 + 5) should cover it in `tests/pipeline_reorg/galaxy_deepseek_prefill_tests.yaml`.

**Review by roseli-TT — APPROVED** [review]
